actors/builtin/multisig/multisig_actor.go

```diff
@@ -141,44 +141,62 @@ func (a *MultiSigActor) Cancel(rt vmr.Runtime, params *TxnIDParams) {
 	// TODO implement cancel logic
 }
 
-type ModifyAuthorizedPartyParams struct {
+type AddAuthorizedParty struct {
 	AuthorizedParty addr.Address // must be an ID protocol address.
+	Increase        bool
 }
 
-func (a *MultiSigActor) AddAuthorizedParty(rt vmr.Runtime, params *ModifyAuthorizedPartyParams) {
+func (a *MultiSigActor) AddAuthorizedParty(rt vmr.Runtime, params *AddAuthorizedParty) {
 	// Can only be called by the multisig wallet itself.
 	rt.ValidateImmediateCallerIs(rt.CurrReceiver())
 
-	actorID, err := addr.IDFromAddress(params.AuthorizedParty)
+	party, err := addr.IDFromAddress(params.AuthorizedParty)
 	if err != nil {
 		rt.AbortStateMsg(fmt.Sprintf("party address is not ID protocol address: %v", err))
 	}
 
 	h, st := a.State(rt)
-	st.AuthorizedParties[abi.ActorID(actorID)] = true
+	if st.isAuthorizedParty(abi.ActorID(party)) {
+		rt.AbortStateMsg("Party is already authorized")
+	}
+	st.AuthorizedParties = append(st.AuthorizedParties, abi.ActorID(party))
+	if params.Increase {
+		st.NumApprovalsThreshold = st.NumApprovalsThreshold + 1
+	}
+
 	UpdateRelease_MultiSig(rt, h, st)
 }
 
-func (a *MultiSigActor) RemoveAuthorizedParty(rt vmr.Runtime, params *ModifyAuthorizedPartyParams) {
+type RemoveAuthorizedParty struct {
+	AuthorizedParty addr.Address // must be an ID protocol address.
+	Decrease        bool
+}
+
+func (a *MultiSigActor) RemoveAuthorizedParty(rt vmr.Runtime, params *RemoveAuthorizedParty) {
 	// Can only be called by the multisig wallet itself.
 	rt.ValidateImmediateCallerIs(rt.CurrReceiver())
 
-	actorID, err := addr.IDFromAddress(params.AuthorizedParty)
+	party, err := addr.IDFromAddress(params.AuthorizedParty)
 	if err != nil {
 		rt.AbortStateMsg(fmt.Sprintf("party address is not ID protocol address: %v", err))
 	}
 
 	h, st := a.State(rt)
 
-	if _, found := st.AuthorizedParties[abi.ActorID(actorID)]; !found {
+	if !st.isAuthorizedParty(abi.ActorID(party)) {
 		rt.AbortStateMsg("Party not found")
 	}
 
-	delete(st.AuthorizedParties, abi.ActorID(actorID))
-
-	if int64(len(st.AuthorizedParties)) < st.NumApprovalsThreshold {
-		rt.AbortStateMsg("Cannot decrease authorized parties below threshold")
+	newAuthorizedParty := make([]abi.ActorID, 0, len(st.AuthorizedParties))
+	for _, s := range st.AuthorizedParties {
+		if s != abi.ActorID(party) {
+			newAuthorizedParty = append(newAuthorizedParty, s)
+		}
+	}
+	if params.Decrease || int64(len(st.AuthorizedParties)-1) < st.NumApprovalsThreshold {
+		st.NumApprovalsThreshold = st.NumApprovalsThreshold - 1
 	}
+	st.AuthorizedParties = newAuthorizedParty
```

Comment:
I have lifted this logic from [lotus](https://github.com/filecoin-project/lotus/blob/master/chain/actors/actor_multisig.go#L358). Flagging as I am unsure what the intended behavior is.

Reply:
i think this is to prevent deadlocking the multisig when the number of authorized parties falls below the threshold upon removal